In [1]:
import sys
import os
sys.path.append(os.path.abspath(os.path.join(os.getcwd(), "../")))

from database_operations import DatabaseOperations
from overtake_model import OvertakingModel
from race_data import RaceDataSetup
from race_dataframe import RaceDataframe
from race_sim import RaceSimulator
from evaluation import RaceSimEvaluation, EvaluateMany

import pandas as pd
pd.set_option("display.max_rows", None)
pd.set_option("display.max_columns", None)
pd.set_option('display.max_colwidth', None) 

In [2]:
db1 = DatabaseOperations(2024, "Mexico City")
race1 = RaceDataframe(db1)
race_data1 = RaceDataSetup(db1, race1)
overtake1 = OvertakingModel(race1.race_df)
test = overtake1.get_model_accuracy()[1]

In [3]:
# race1.race_df[race1.race_df["lap_num"]>69]

In [4]:
# print(race_data1.driver_tyre_coefficients)

race_sim = RaceSimulator(race_data1, overtake1)
race_sim.simulate()
print(race_sim.num_overtakes)
sim_df = race_sim.get_results_as_dataframe()
sim_df

77


driver_number       driver_name                 pit_schedule  tyre_type  \
0              55      Carlos Sainz                {33: 3, 1: 2}          3   
1               4      Lando Norris                {31: 3, 1: 2}          3   
2              16   Charles Leclerc         {32: 3, 70: 1, 1: 2}          1   
3              44    Lewis Hamilton                {29: 3, 1: 2}          3   
4              63    George Russell                {32: 3, 1: 2}          3   
5               1    Max Verstappen                {27: 3, 1: 2}          3   
6              81     Oscar Piastri                {40: 3, 1: 2}          3   
7              20   Kevin Magnussen                {31: 3, 1: 2}          3   
8              27   Nico Hulkenberg                {30: 3, 1: 2}          3   
9              10      Pierre Gasly                {29: 3, 1: 2}          3   
10             45  Franco Colapinto                {48: 2, 1: 3}          2   
11             18      Lance Stroll                {27: 3, 1: 2}          3   
12             31      Esteban Ocon                {1: 3, 49: 2}          2   
13             40       Liam Lawson         {40: 2, 66: 1, 1: 3}          1   
14             77   Valtteri Bottas                {50: 2, 1: 3}          2   
15             24       Guanyu Zhou                {44: 2, 1: 3}          2   
16             11      Sergio Perez  {21: 2, 44: 2, 69: 1, 1: 3}          1   
17             23   Alexander Albon                       {1: 2}          2   
18             22      Yuki Tsunoda                       {1: 2}          2   
19             14   Fernando Alonso                       {1: 2}          2   

    lap_num  sector  sector_time  stint_lap  cumulative_time        gap  \
0        71       3    20.621724         39      6262.958798   0.000000   
1        71       3    20.595324         41      6263.862916   0.904118   
2        71       3    21.478718          2      6284.163533  20.300617   
3        71       3    20.863496         43      6298.256638  14.093105   
4        71       3    20.987235         40      6310.060496  11.803859   
5        71       3    20.817359         45      6312.124494   2.063998   
6        71       3    20.743440         32      6313.372574   1.248080   
7        71       3    21.039990         41      6323.509212  10.136638   
8        71       3    20.855889         42      6337.618340  14.109128   
9        71       3    21.015707         43      6356.214422  18.596082   
10       71       3    20.918221         24      6360.944877   4.730455   
11       71       3    20.813953         45      6361.776555   0.831678   
12       71       3    21.052468         23      6378.444910  16.668355   
13       71       3    22.196024          6      6383.276963   4.832053   
14       71       3    21.274395         22      6388.012064   4.735101   
15       71       3    21.132479         28      6398.351809  10.339745   
16       71       3    21.347959          3      6415.532899  17.181090   
17       71       0     0.000000         71         0.000000   0.000000   
18       71       0     0.000000         71         0.000000   0.000000   
19       71       3    23.298222         71      1489.244362   2.823258   

      pit       pace  position  starting_pos  \
0   False  28.565273         1             1   
1   False  20.693255         2             2   
2   False  21.183077         3             3   
3   False  20.932366         4             4   
4   False  21.022015         5             5   
5   False  20.890227         6             6   
6   False  20.795347         7             8   
7   False  21.082029         8             7   
8   False  20.934665         9             9   
9   False  21.082268        10            10   
10  False  21.043121        11            12   
11  False  20.893413        12            11   
12  False  21.090358        13            13   
13  False  21.273599        14            16   
14  False  21.222770        15            14   


In [5]:
eval = RaceSimEvaluation(race_sim, race1, db1)

print(eval.compare_total_cumulative_times()[1])
eval.compare_total_cumulative_times()[0]


269.67138129196815


driver_number       driver_name  simulated_cumulative_time  \
0              55      Carlos Sainz                6262.958798   
1               4      Lando Norris                6263.862916   
2              16   Charles Leclerc                6284.163533   
3              44    Lewis Hamilton                6298.256638   
4              63    George Russell                6310.060496   
5               1    Max Verstappen                6312.124494   
6              81     Oscar Piastri                6313.372574   
7              20   Kevin Magnussen                6323.509212   
8              27   Nico Hulkenberg                6337.618340   
9              10      Pierre Gasly                6356.214422   
10             45  Franco Colapinto                6360.944877   
11             18      Lance Stroll                6361.776555   
12             31      Esteban Ocon                6378.444910   
13             40       Liam Lawson                6383.276963   
14             77   Valtteri Bottas                6388.012064   
15             24       Guanyu Zhou                6398.351809   
16             11      Sergio Perez                6415.532899   
17             23   Alexander Albon                   0.000000   
18             22      Yuki Tsunoda                   0.000000   
19             14   Fernando Alonso                1489.244362   

    actual_cumulative_time  absolute_error  
0                 6023.912      239.046798  
1                 6028.139      235.723916  
2                 6057.475      226.688533  
3                 6067.579      230.677638  
4                 6071.151      238.909496  
5                 6083.463      228.661494  
6                 6083.971      229.401574  
7                 6086.111      237.398212  
8                 6024.711      312.907340  
9                 5959.340      396.874422  
10                6047.376      313.568877  
11                6045.905      315.871555  
12                6053.418      325.026910  
13                6085.804      297.472963  
14                6062.091      325.921064  
15                6081.243      317.108809  
16                6093.613      321.919899  
17                     NaN             NaN  
18                     NaN             NaN  
19                1428.339       60.905362

In [6]:
eval.get_position_accuracy_final_class()

{'position_accuracy': 0.5882352941176471,
 'top_3_accuracy': 1.0,
 'mean_error': 0.47058823529411764,
 'total_error': 8}

In [7]:
eval.compare_total_to_front()

{'total_simulated_gap_to_front': 1278.1819409712125,
 'total_actual_gap_to_front': 1272.1519999999973}

{'total_simulated_gap_to_front': 1305.9637926282421,
 'total_actual_gap_to_front': 1272.1519999999973}

In [8]:
from typing import List, Dict

def find_most_accurate_races(race_inputs: List[Dict]) -> List[Dict]:
	"""
	Finds the 4 most accurately simulated races based on the average mean error.

	Args:
		race_inputs (List[Dict]): A list of dictionaries, where each dictionary contains the inputs
								  for a race simulation (year, circuit, etc.).

	Returns:
		List[Dict]: A list of the 4 most accurate races, sorted by their average mean error (lowest first).
					Each dictionary includes the race inputs and its evaluation metrics.
	"""
	race_evaluations = []

	# Iterate through each race input
	for race_input in race_inputs:
		year = race_input["year"]
		circuit = race_input["circuit"]

		try:
			# Initialize database, race data, setup, overtaking model, and simulation
			db = DatabaseOperations(year, circuit)
			race_data = RaceDataframe(db)
			race_setup = RaceDataSetup(db, race_data)
			overtake = OvertakingModel(race_data.race_df)
			simulation = RaceSimulator(race_setup, overtake)
			simulation.simulate()

			# Evaluate the simulation
			eval = RaceSimEvaluation(simulation, race_data, db)
			metrics = eval.get_position_accuracy_final_class()
			# Store the race inputs and evaluation metrics
			race_evaluations.append({
				"race_input": race_input,
				"metrics": metrics,
				"overtakes": simulation.num_overtakes,
				"has_safety_car": simulation.has_safety_car
			})

		except Exception as e:
			# Log the error and continue with the next race
			print(f"Error evaluating race {year} at {circuit}: {e}")
			continue

	# Sort the races by average mean error (lower is better)
	sorted_races = sorted(
		race_evaluations,
		key=lambda x: x["metrics"]["mean_error"]
	)

	# Return the top 4 most accurate races
	return sorted_races